app.go

```diff
@@ -757,20 +774,46 @@ func (app *App) start(ctx context.Context) error {
 
 			return multierr.Append(err, stopErr)
 		}
+
 		return err
 	}
 	app.log.LogEvent(&fxevent.Running{})
 
 	return nil
 }
 
-func withTimeout(ctx context.Context, f func(context.Context) error) error {
+type withTimeoutParams struct {
+	log       fxevent.Logger
+	hook      string
+	ctx       context.Context
+	callback  func(context.Context) error
+	lifecycle *lifecycleWrapper
+}
+
+func withTimeout(param *withTimeoutParams) error {
 	c := make(chan error, 1)
-	go func() { c <- f(ctx) }()
+	ctx := param.ctx
+	go func() { c <- param.callback(ctx) }()
 
 	select {
 	case <-ctx.Done():
-		return ctx.Err()
+		err := ctx.Err()
+		if err != context.DeadlineExceeded {
+			return err
+		}
+		// On timeout, report running hook's caller and recorded
+		// runtimes of hooks successfully run till end.
+		r := param.lifecycle.hookRecords()
+		caller := param.lifecycle.runningHookCaller()
+		if len(r) > 0 {
+			err = multierr.Append(err, fmt.Errorf(`
+timed out while executing hook %s (Caller: %s). Hooks successfully ran so far: %s`,
+				param.hook, caller, r))
+		} else {
+			err = multierr.Append(err, fmt.Errorf(`
+timed out while executing hook %s (Caller: %s)`, param.hook, caller))
```

Comment:
nit about the message: how about something like,

    "%v hook added by %v failed: %w\n%s", param.hook, caller, err, r

So we get something like:

    OnStart hook added by whatever.func() failed: deadline exceeded
    [hook run report here]

(`%w` above is equivalent to creating a wrapper; if you create a wrapper error type instead, that `%w` can be a `%v` in the `fmt.Sprintf` call.)

Maybe we can also include the name of the function that failed to run? I don't know the exact framing of the message for that, though.